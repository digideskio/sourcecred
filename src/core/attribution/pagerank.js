// @flow

import {type Edge, Graph} from "../graph";
import {
  distributionToNodeDistribution,
  createContributions,
  createOrderedSparseMarkovChain,
  type EdgeWeight,
} from "./graphToMarkovChain";
import {
  decompose,
  type PagerankNodeDecomposition,
} from "./pagerankNodeDecomposition";

import {findStationaryDistribution} from "./markovChain";

export type {NodeDistribution} from "./graphToMarkovChain";
export type {PagerankNodeDecomposition} from "./pagerankNodeDecomposition";
export type PagerankOptions = {|
  +selfLoopWeight?: number,
  +verbose?: boolean,
  +convergenceThreshold?: number,
  +maxIterations?: number,
|};

export type {EdgeWeight} from "./graphToMarkovChain";
export type EdgeEvaluator = (Edge) => EdgeWeight;

function defaultOptions(): PagerankOptions {
  return {
    verbose: false,
    selfLoopWeight: 1e-3,
    convergenceThreshold: 1e-7,
    maxIterations: 255,
  };
}

export function pagerank(
  graph: Graph,
  edgeWeight: EdgeEvaluator,
  options?: PagerankOptions
): PagerankNodeDecomposition {
  const fullOptions = {
    ...defaultOptions(),
    ...(options || {}),
  };
  const contributions = createContributions(
    graph,
    edgeWeight,
    fullOptions.selfLoopWeight
  );
  const osmc = createOrderedSparseMarkovChain(contributions);
  const distribution = findStationaryDistribution(osmc.chain, {
    verbose: fullOptions.verbose,
    convergenceThreshold: fullOptions.convergenceThreshold,
    maxIterations: fullOptions.maxIterations,
  });
  const pi = distributionToNodeDistribution(osmc.nodeOrder, distribution);
  return decompose(pi, contributions);
}
