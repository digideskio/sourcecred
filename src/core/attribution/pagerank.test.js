// @flow

import {pagerank} from "./pagerank";
import {EdgeAddress, Graph, NodeAddress, edgeToStrings} from "../graph";
import {advancedGraph} from "../graphTestUtil";
import * as MapUtil from "../../util/map";

/**
 * Format a decomposition to be shown in a snapshot. This converts
 * addresses and edges to strings to avoid NUL characters.
 */
function formatDecomposition(d) {
  return MapUtil.mapEntries(d, (key, {score, scoredContributions}) => [
    NodeAddress.toString(key),
    {
      score,
      scoredContributions: scoredContributions.map(
        ({contribution, source, sourceScore, contributionScore}) => ({
          contribution: {
            contributor: formatContributor(contribution.contributor),
            weight: contribution.weight,
          },
          source: NodeAddress.toString(source),
          sourceScore,
          contributionScore,
        })
      ),
    },
  ]);
  function formatContributor(contributor) {
    switch (contributor.type) {
      case "SYNTHETIC_LOOP":
        return {type: "SYNTHETIC_LOOP"};
      case "IN_EDGE":
        return {type: "IN_EDGE", edge: edgeToStrings(contributor.edge)};
      case "OUT_EDGE":
        return {type: "OUT_EDGE", edge: edgeToStrings(contributor.edge)};
      default:
        throw new Error((contributor.type: empty));
    }
  }
}

function snapshotPagerankResult(result) {
  expect(formatDecomposition(result)).toMatchSnapshot();
}

describe("core/attribution/pagerank", () => {
  function edgeWeight(_unused_edge) {
    return {toWeight: 1, froWeight: 0};
  }
  it("snapshots as expected on the advanced graph", () => {
    const pagerankResult = pagerank(advancedGraph().graph1(), edgeWeight);
    snapshotPagerankResult(pagerankResult);
  });
  it("respects explicit arguments", () => {
    const pagerankResult = pagerank(advancedGraph().graph1(), edgeWeight, {
      maxIterations: 0,
    });
    snapshotPagerankResult(pagerankResult);
  });
});
