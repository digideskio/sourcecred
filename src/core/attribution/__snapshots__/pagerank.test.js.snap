// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/attribution/pagerank respects explicit arguments 1`] = `
Map {
  "NodeAddress[\\"src\\"]" => Object {
    "score": 0.25,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 0.0004997501249375313,
        },
        "contributionScore": 0.00012493753123438283,
        "source": "NodeAddress[\\"src\\"]",
        "sourceScore": 0.25,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"1\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "OUT_EDGE",
          },
          "weight": 0,
        },
        "contributionScore": 0,
        "source": "NodeAddress[\\"dst\\"]",
        "sourceScore": 0.25,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"2\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "OUT_EDGE",
          },
          "weight": 0,
        },
        "contributionScore": 0,
        "source": "NodeAddress[\\"dst\\"]",
        "sourceScore": 0.25,
      },
    ],
  },
  "NodeAddress[\\"dst\\"]" => Object {
    "score": 0.25,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 1,
        },
        "contributionScore": 0.25,
        "source": "NodeAddress[\\"dst\\"]",
        "sourceScore": 0.25,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"1\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "IN_EDGE",
          },
          "weight": 0.49975012493753124,
        },
        "contributionScore": 0.12493753123438281,
        "source": "NodeAddress[\\"src\\"]",
        "sourceScore": 0.25,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"2\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "IN_EDGE",
          },
          "weight": 0.49975012493753124,
        },
        "contributionScore": 0.12493753123438281,
        "source": "NodeAddress[\\"src\\"]",
        "sourceScore": 0.25,
      },
    ],
  },
  "NodeAddress[\\"loop\\"]" => Object {
    "score": 0.25,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"loop\\"]",
              "dst": "NodeAddress[\\"loop\\"]",
              "src": "NodeAddress[\\"loop\\"]",
            },
            "type": "IN_EDGE",
          },
          "weight": 0.9990009990009991,
        },
        "contributionScore": 0.2497502497502498,
        "source": "NodeAddress[\\"loop\\"]",
        "sourceScore": 0.25,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 0.0009990009990009992,
        },
        "contributionScore": 0.0002497502497502498,
        "source": "NodeAddress[\\"loop\\"]",
        "sourceScore": 0.25,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"loop\\"]",
              "dst": "NodeAddress[\\"loop\\"]",
              "src": "NodeAddress[\\"loop\\"]",
            },
            "type": "OUT_EDGE",
          },
          "weight": 0,
        },
        "contributionScore": 0,
        "source": "NodeAddress[\\"loop\\"]",
        "sourceScore": 0.25,
      },
    ],
  },
  "NodeAddress[\\"isolated\\"]" => Object {
    "score": 0.25,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 1,
        },
        "contributionScore": 0.25,
        "source": "NodeAddress[\\"isolated\\"]",
        "sourceScore": 0.25,
      },
    ],
  },
}
`;

exports[`core/attribution/pagerank snapshots as expected on the advanced graph 1`] = `
Map {
  "NodeAddress[\\"src\\"]" => Object {
    "score": 3.120317183596679e-11,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 0.0004997501249375313,
        },
        "contributionScore": 1.5593789023471662e-14,
        "source": "NodeAddress[\\"src\\"]",
        "sourceScore": 3.120317183596679e-11,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"1\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "OUT_EDGE",
          },
          "weight": 0,
        },
        "contributionScore": 0,
        "source": "NodeAddress[\\"dst\\"]",
        "sourceScore": 0.4999999999687968,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"2\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "OUT_EDGE",
          },
          "weight": 0,
        },
        "contributionScore": 0,
        "source": "NodeAddress[\\"dst\\"]",
        "sourceScore": 0.4999999999687968,
      },
    ],
  },
  "NodeAddress[\\"dst\\"]" => Object {
    "score": 0.4999999999687968,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 1,
        },
        "contributionScore": 0.4999999999687968,
        "source": "NodeAddress[\\"dst\\"]",
        "sourceScore": 0.4999999999687968,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"1\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "IN_EDGE",
          },
          "weight": 0.49975012493753124,
        },
        "contributionScore": 1.5593789023471658e-11,
        "source": "NodeAddress[\\"src\\"]",
        "sourceScore": 3.120317183596679e-11,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"hom\\",\\"2\\"]",
              "dst": "NodeAddress[\\"dst\\"]",
              "src": "NodeAddress[\\"src\\"]",
            },
            "type": "IN_EDGE",
          },
          "weight": 0.49975012493753124,
        },
        "contributionScore": 1.5593789023471658e-11,
        "source": "NodeAddress[\\"src\\"]",
        "sourceScore": 3.120317183596679e-11,
      },
    ],
  },
  "NodeAddress[\\"loop\\"]" => Object {
    "score": 0.25000000000000017,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"loop\\"]",
              "dst": "NodeAddress[\\"loop\\"]",
              "src": "NodeAddress[\\"loop\\"]",
            },
            "type": "IN_EDGE",
          },
          "weight": 0.9990009990009991,
        },
        "contributionScore": 0.24975024975024995,
        "source": "NodeAddress[\\"loop\\"]",
        "sourceScore": 0.25000000000000017,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 0.0009990009990009992,
        },
        "contributionScore": 0.00024975024975024997,
        "source": "NodeAddress[\\"loop\\"]",
        "sourceScore": 0.25000000000000017,
      },
      Object {
        "contribution": Object {
          "contributor": Object {
            "edge": Object {
              "address": "EdgeAddress[\\"loop\\"]",
              "dst": "NodeAddress[\\"loop\\"]",
              "src": "NodeAddress[\\"loop\\"]",
            },
            "type": "OUT_EDGE",
          },
          "weight": 0,
        },
        "contributionScore": 0,
        "source": "NodeAddress[\\"loop\\"]",
        "sourceScore": 0.25000000000000017,
      },
    ],
  },
  "NodeAddress[\\"isolated\\"]" => Object {
    "score": 0.25,
    "scoredContributions": Array [
      Object {
        "contribution": Object {
          "contributor": Object {
            "type": "SYNTHETIC_LOOP",
          },
          "weight": 1,
        },
        "contributionScore": 0.25,
        "source": "NodeAddress[\\"isolated\\"]",
        "sourceScore": 0.25,
      },
    ],
  },
}
`;
